reporting_app/static/library_preps.js

```diff
@@ -0,0 +1,201 @@
+    // e.g, query_nested_object({'x': {'y': 'z'}}, ['x', 'y']) -> 'z'
+    var val = top_level;
+    for (var i=0; i<query.length; i++) {
+        val = val[query[i]];
+        if (!val) {
+            return null;
+        }
+    }
+    return val;
+}
+
+function build_series(colour_metric, format_func) {
+    // build an object to pass to Highcharts.heatmap.series, converting library QC data into an array of data points
+    var series = {
+        name: library_data['id'],
+        data: [],
+        dataLabels: {enabled: false}
+    }
+
+    var placements = library_data['qc'];
+    for (coord in placements) {
+        var split_coord = coord.split(':');
+        series.data.push(
+            {
+                y: heatmap_x_coords.indexOf(split_coord[0]),
+                x: parseInt(split_coord[1]) - 1,
+                value: format_func(query_nested_object(placements[coord], metrics[colour_metric]['path'])),
+                name: placements[coord]['name']
+            }
+        );
+    }
+    return series;
+}
+
+function highchart(heatmap_id, title, lims_url, qc_url, ajax_token, library_id) {
+    chart = Highcharts.chart(heatmap_id, {
+        chart: {
+            type: 'heatmap',
+            marginTop: 40,
+            marginBottom: 80,
+            plotBorderWidth: 1,
+            width: 800,
+            height: 400
+        },
+        title: {text: 'Library ' + title},
+        yAxis: {categories: heatmap_x_coords, max: 7, reversed: true, title: null},
+        xAxis: {categories: ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12'], max: 11},
+        colorAxis: {
+            minColor: '#FFFFFF',
+            maxColor: colour_palette[0]
+        },
+        legend: {
+            title: {
+                text: metrics[active_colour_metric]['name']
+            },
+            align: 'right',
+            layout: 'vertical',
+            margin: 0,
+            verticalAlign: 'top',
+            y: 25
+        },
+        tooltip: {
+            formatter: function () {
+                var coord = this.series.yAxis.categories[this.point.y] + ':' + (this.point.x + 1);
+                return '<p>Sample: ' + this.point.name + '<br/>Coord: ' + coord + '<br/>' + metrics[active_colour_metric]['name'] + ': ' + this.point.value + '</p>';
+            }
+        },
+        series: []
+    });
+    chart.showLoading();
+    get_lims_and_qc_data(lims_url, qc_url, ajax_token, library_id);
+    chart.hideLoading();
+}
+
+function data_classes(categories) {
+    // Map a list of category names into a list of dataClasses to pass to Highcharts.heatmap.colorAxis.dataClasses
+    var data_classes = [];
+    for (var i=0; i<categories.length; i++) {
+        data_classes.push(
+            {
+                from: i,
+                to: i + 1,
+                color: colour_palette[i],
+                name: categories[i]
+            }
+        );
+    }
+    return data_classes;
+}
+
+function trigger(colour_metric) {
```

Comment:
Same as above. Unless we encapsulate the function in object we need them to be a bit more specific.